src/main.cc

```diff
@@ -214,6 +231,145 @@ int main(int argc, char** argv, char** envp)
+#endif
+                log_error("Unable to change user.");
+                exit(EXIT_FAILURE);
+            }
+            log_info("Dropped to User: {}", user->c_str());
+        }
+
+        // are we requested to daemonize?
+        bool daemon = opts["daemon"].as<bool>();
+        if (daemon) {
+            pid_t pid;
+
+            // fork
+            pid = fork();
+            if (pid < 0) {
+                log_error("Unable to fork.");
+                exit(EXIT_FAILURE);
+            }
+
+            // terminate parent
+            if (pid > 0)
+                exit(EXIT_SUCCESS);
+
+            // become session leader
+            if (setsid() < 0) {
+                log_error("Unable to setsid.");
+                exit(EXIT_FAILURE);
+            }
+
+            // second fork
+            pid = fork();
+            if (pid < 0) {
+                log_error("Unable to fork.");
+                exit(EXIT_FAILURE);
+            }
+
+            // terminate parent
+            if (pid > 0)
+                exit(EXIT_SUCCESS);
+
+            // set new file permissions
+            umask(0);
+
+            // change dir to /
+            if (chdir("/")) {
+                log_error("Unable to chdir to root dir.");
+                exit(EXIT_FAILURE);
+            }
+            // close open filedescriptors belonging to a tty
+            for (int fd = sysconf(_SC_OPEN_MAX); fd >= 0; fd--) {
+                if (isatty(fd))
+                    close(fd);
+            }
+            log_info("Daemonized.");
+        }
+
+        // are we requested to write a pidfile ?
+        std::optional<std::string> pidfile;
+        if (opts.count("pidfile") > 0) {
+            pidfile = opts["pidfile"].as<std::string>();
+
+            // exit if the pidfile already exists
+            struct stat* s = NULL;
+            errno = 0;
+            // for pidfile to be not there stat needs to fail and it needs
+            // to fail due to ENOENT (no such file or dir)
+            if (!(-1 == stat(pidfile->c_str(), s) && errno == ENOENT)) {
+                log_error("Pidfile {} exists. It may be that gerbera is already", pidfile->c_str());
+                log_error("running or the file is a leftover from an unclean shutdown.");
+                log_error("In that case, remove the file before starting gerbera.");
+                exit(EXIT_FAILURE);
+            }
+
+            // get the pid of our process
+            pid_t pid = getpid();
+
+            // convert to a string
+            char pidstr[20];
+            if (0 > snprintf(pidstr, 18, "%d", pid)) {
+                log_error("Could not determine pid of running process.");
+                exit(EXIT_FAILURE);
+            }
+
+            // add a newline
+            strcat(pidstr, "\n");
+
+            // open the pidfile
+            int pidfd = open(pidfile->c_str(), O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
```

Comment:
There is a function `writeTextFile` in `util/tools` which does the file work for you.

Reply:
I will look into it within the next week.